File: src/main/java/org/docopt/LeafPattern.java
package org.docopt;

import java.util.List;

/**
 * Leaf/terminal node of a pattern tree.
 */
abstract class LeafPattern extends Pattern {

    private final String name;

    private Object value;

    public LeafPattern(final String name, final Object value) {
        this.name = name;
        this.value = value;
    }

    public LeafPattern(final String name) {
        this(name, null);
    }

    @Override
    public String toString() {
        return String.format("%s(%s, %s)", getClass().getSimpleName(),
                Py.INSTANCE.repr(name), Py.INSTANCE.repr(value));
    }

    @Override
    protected final List<Pattern> flat(final Class<?>... types) {
        {

            if (!Py.INSTANCE.bool(types) || Py.INSTANCE.in(getClass(), types)) {
                return Py.INSTANCE.list((Pattern) this);
            }

            return Py.INSTANCE.list();
        }
    }

    @Override
    protected MatchResult match(final List<LeafPattern> left,
                                List<LeafPattern> collected) {
        if (collected == null) {
            collected = Py.INSTANCE.list();
        }

        Integer pos;
        LeafPattern match;

        {
            final SingleMatchResult m = singleMatch(left);
            pos = m.getPosition();
            match = m.getMatch();
        }

        if (match == null) {
            return new MatchResult(false, left, collected);
        }

        List<LeafPattern> left_;

        {
            left_ = Py.INSTANCE.list();
            left_.addAll(left.subList(0, pos));

            if ((pos + 1) < left.size()) {
                left_.addAll(left.subList(pos + 1, left.size()));
            }
        }

        List<LeafPattern> sameName;

        {
            sameName = Py.INSTANCE.list();

            for (final LeafPattern a : collected) {
                if (name.equals(a.getName())) {
                    sameName.add(a);
                }
            }
        }

        Object increment;

        if ((value instanceof Integer) || (value instanceof List)) {
            if (value instanceof Integer) {
                increment = 1;
            } else {
                final Object v = match.getValue();
                increment = (v instanceof String) ? Py.INSTANCE.list(v) : v;
            }

            if (sameName.isEmpty()) {
                match.setValue(increment);
                return new MatchResult(true, left_,
                        Py.INSTANCE.plus(collected, Py.INSTANCE.list(match)));
            }

            {
                final LeafPattern p = sameName.get(0);
                final Object v = p.getValue();

                if (v instanceof Integer) {
                    final Integer a = (Integer) v;
                    final Integer b = (Integer) increment;
                    p.setValue(a + b);
                } else if (v instanceof List) {
                    @SuppressWarnings("unchecked") final List<LeafPattern> a = (List<LeafPattern>) v;
                    @SuppressWarnings("unchecked") final List<LeafPattern> b = (List<LeafPattern>) increment;
                    a.addAll(b);
                }
            }

            // TODO: Should collected be copied to a new list?
            return new MatchResult(true, left_, collected);
        }
        return new MatchResult(true, left_, Py.INSTANCE.plus(collected, Py.INSTANCE.list(match)));
    }

    protected abstract SingleMatchResult singleMatch(List<LeafPattern> left);

    public String getName() {
        return name;
    }

    public Object getValue() {
        return value;
    }

    public void setValue(final Object value) {
        this.value = value;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = super.hashCode();
        result = prime * result + ((name == null) ? 0 : name.hashCode());
        result = prime * result + ((value == null) ? 0 : value.hashCode());
        return result;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final LeafPattern other = (LeafPattern) obj;
        if (name == null) {
            if (other.name != null) {
                return false;
            }
        } else if (!name.equals(other.name)) {
            return false;
        }
        if (value == null) {
            return other.value == null;
        } else return value.equals(other.value);
    }
}

File: src/main/java/org/docopt/LeafPattern.kt
package org.docopt

import org.docopt.Py.`in`
import org.docopt.Py.bool
import org.docopt.Py.list
import org.docopt.Py.plus
import org.docopt.Py.repr

/**
 * Leaf/terminal node of a pattern tree.
 */
internal abstract class LeafPattern @JvmOverloads constructor(
    val name: String?,
    var value: Any? = null
) : Pattern() {
    override fun toString(): String {
        return String.format(
            "%s(%s, %s)", javaClass.simpleName,
            repr(name), repr(value)
        )
    }

    override fun flat(vararg types: Class<*>): List<Pattern> {
        run {
            return if (!bool(types) || `in`<Class<*>>(javaClass, *types)) {
                list(this as Pattern)
            } else list()
        }
    }

    override fun match(
        left: List<LeafPattern>,
        collected: List<LeafPattern>
    ): MatchResult {
        var collected: List<LeafPattern>? = collected
        if (collected == null) {
            collected = list()
        }
        var pos: Int?
        var match: LeafPattern?
        run {
            val m = singleMatch(left)
            pos = m.position
            match = m.match
        }
        if (match == null) {
            return MatchResult(false, left, collected)
        }
        var left_: MutableList<LeafPattern>
        run {
            left_ = list()
            left_.addAll(left.subList(0, pos!!))
            if (pos!! + 1 < left.size) {
                left_.addAll(left.subList(pos!! + 1, left.size))
            }
        }
        var sameName: MutableList<LeafPattern>
        run {
            sameName = list()
            for (a in collected) {
                if (name == a.name) {
                    sameName.add(a)
                }
            }
        }
        val increment: Any
        if (value is Int || value is List<*>) {
            increment = if (value is Int) {
                1
            } else {
                val v = match!!.value
                (if (v is String) list<Any>(v) else v)!!
            }
            if (sameName.isEmpty()) {
                match!!.value = increment
                return MatchResult(
                    true, left_,
                    plus(collected, list(match!!))
                )
            }
            run {
                val p = sameName[0]
                val v = p.value
                if (v is Int) {
                    val b = increment as Int
                    p.value = v + b
                } else if (v is List<*>) {
                    val a = v as MutableList<LeafPattern>
                    val b = increment as List<LeafPattern>
                    a.addAll(b)
                }
            }

            // TODO: Should collected be copied to a new list?
            return MatchResult(true, left_, collected)
        }
        return MatchResult(
            true, left_, plus(
                collected, list(
                    match!!
                )
            )
        )
    }

    protected abstract fun singleMatch(left: List<LeafPattern>): SingleMatchResult
    override fun hashCode(): Int {
        val prime = 31
        var result = super.hashCode()
        result = prime * result + (name?.hashCode() ?: 0)
        result = prime * result + if (value == null) 0 else value.hashCode()
        return result
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) {
            return true
        }
        if (javaClass != obj!!.javaClass) {
            return false
        }
        val other = obj as LeafPattern?
        if (name == null) {
            if (other!!.name != null) {
                return false
            }
        } else if (name != other!!.name) {
            return false
        }
        return if (value == null) {
            other.value == null
        } else value == other.value
    }
}